src/Ocelot.Provider.Consul/ConsulProviderFactory.cs

```diff
@@ -1,41 +1,41 @@
-﻿using Microsoft.Extensions.DependencyInjection;
+﻿using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.DependencyInjection;
 using Ocelot.Configuration;
 using Ocelot.Logging;
 using Ocelot.Provider.Consul.Interfaces;
 using Ocelot.ServiceDiscovery.Providers;
 
 namespace Ocelot.Provider.Consul;
 
-public static class ConsulProviderFactory
+/// <summary>
+/// TODO It must be refactored converting to real factory-class and add to DI.
+/// </summary>
+/// <remarks>
+/// Must inherit from <see cref="IServiceDiscoveryProviderFactory"/> interface.
+/// Also the <see cref="ServiceDiscoveryFinderDelegate"/> must be removed from the design.
+/// </remarks>
+public static class ConsulProviderFactory // TODO : IServiceDiscoveryProviderFactory
 {
-    /// <summary>
-    /// String constant used for provider type definition.
-    /// </summary>
+    /// <summary>String constant used for provider type definition.</summary>
     public const string PollConsul = nameof(Provider.Consul.PollConsul);
 
-    private static readonly List<PollConsul> ServiceDiscoveryProviders = new();
-    private static readonly object LockObject = new();
+    private static readonly List<PollConsul> ServiceDiscoveryProviders = new(); // TODO It must be scoped service in DI-container
+    private static readonly object SyncRoot = new();
 
     public static ServiceDiscoveryFinderDelegate Get { get; } = CreateProvider;
-
-    private static ConsulRegistryConfiguration configuration;
-    private static ConsulRegistryConfiguration ConfigurationGetter() => configuration;
-    public static Func<ConsulRegistryConfiguration> GetConfiguration { get; } = ConfigurationGetter;
-
-    private static IServiceDiscoveryProvider CreateProvider(IServiceProvider provider,
-        ServiceProviderConfiguration config, DownstreamRoute route)
+    private static IServiceDiscoveryProvider CreateProvider(IServiceProvider provider, ServiceProviderConfiguration config, DownstreamRoute route)
     {
         var factory = provider.GetService<IOcelotLoggerFactory>();
         var consulFactory = provider.GetService<IConsulClientFactory>();
-
-        configuration = new ConsulRegistryConfiguration(config.Scheme, config.Host, config.Port, route.ServiceName, config.Token);
+        var configuration = new ConsulRegistryConfiguration(config.Scheme, config.Host, config.Port, route.ServiceName, config.Token);
+        var contextAccessor = provider.GetService<IHttpContextAccessor>();
+        contextAccessor.HttpContext.Items[nameof(ConsulRegistryConfiguration)] = configuration;
```

Comment:
Great idea to use the http context items as storage of the consul registry configuration. Just a stupid question. Could we have some side effects with the multiplexer (copies of the http context)?

Reply:
Good question!... `IServiceDiscoveryProviderFactory` is a part of `LoadBalancerFactory` which is called in `LoadBalancerHouse` and which belongs to `LoadBalancingMiddleware`, for sure. So, any service discovery providers belong to `LoadBalancingMiddleware`, including our lovely `Consul` provider.

To answer on your question we have to compare both `LoadBalancingMiddleware` and `MultiplexingMiddleware`, and their order in Ocelot pipeline ->
https://github.com/ThreeMammals/Ocelot/blob/58d87c9273c1252a9da64a8900b080c96511d480/src/Ocelot/Middleware/OcelotPipelineExtensions.cs#L27
https://github.com/ThreeMammals/Ocelot/blob/58d87c9273c1252a9da64a8900b080c96511d480/src/Ocelot/Middleware/OcelotPipelineExtensions.cs#L59
https://github.com/ThreeMammals/Ocelot/blob/58d87c9273c1252a9da64a8900b080c96511d480/src/Ocelot/Middleware/OcelotPipelineExtensions.cs#L134
So... early bird multiplexer calls load-balancing middleware in this method on line 190:
https://github.com/ThreeMammals/Ocelot/blob/58d87c9273c1252a9da64a8900b080c96511d480/src/Ocelot/Multiplexer/MultiplexingMiddleware.cs#L184-L192
And next delegate receives new context object, right? Therefore `LoadBalancingMiddleware` receives new `HttpContext` object. But, seems it doesn't matter we process new or old context object, because of `IHttpContextAccessor` injecting.
The questions should be formulated as follows: "Do both `IHttpContextAccessor` objects in `ConsulProviderFactory` and `DefaultConsulServiceBuilder` operate by the same `HttpContext` objects? And once again it doesn't matter either it's new or old one. And I hope soo. Because we are in the context of load balancing logic we should have the same `HttpContext`.

Moreover I've checked DI container setup and our `IHttpContextAccessor` service is singleton:
https://github.com/ThreeMammals/Ocelot/blob/58d87c9273c1252a9da64a8900b080c96511d480/src/Ocelot/DependencyInjection/OcelotBuilder.cs#L125
Seems we process old context object, not one from multiplexer. 😢 
But anyway, it doesn't matter, and we use the same context object in `ConsulProviderFactory` and `DefaultConsulServiceBuilder`, so we should have any exceptions or empty config objects here while reading:
https://github.com/ThreeMammals/Ocelot/blob/58d87c9273c1252a9da64a8900b080c96511d480/src/Ocelot/DependencyInjection/OcelotBuilder.cs#L125
**Ups**❗ Why is it added as singleton? I thought it is scoped service. 😠

Reply:
Tom added this service as singleton in 2017... 🤯 
- commit https://github.com/ThreeMammals/Ocelot/commit/6289992faab8a4138e0beb0876da9b73c0987f08#diff-8f16fc8879d2a7b30200bf1a3aafbb7baf957f49c53db5db704a3c3fbd985c71R119
- PR #159  

Oh, no!... This problem exists since ages ❗ 🤕

Reply:
My Visual Studio disassembler shows this code:
```c#
    /// <summary>
    /// Provides an implementation of <see cref="IHttpContextAccessor" /> based on the current execution context.
    /// </summary>
    public class HttpContextAccessor : IHttpContextAccessor
    {
        private static readonly AsyncLocal<HttpContextHolder> _httpContextCurrent = new AsyncLocal<HttpContextHolder>();

        /// <inheritdoc/>
        public HttpContext? HttpContext
        {
            get
            {
                return  _httpContextCurrent.Value?.Context;
            }
            set
            {
                var holder = _httpContextCurrent.Value;
                if (holder != null)
                {
                    // Clear current HttpContext trapped in the AsyncLocals, as its done.
                    holder.Context = null;
                }

                if (value != null)
                {
                    // Use an object indirection to hold the HttpContext in the AsyncLocal,
                    // so it can be cleared in all ExecutionContexts when its cleared.
                    _httpContextCurrent.Value = new HttpContextHolder { Context = value };
                }
            }
        }

        private class HttpContextHolder
        {
            public HttpContext? Context;
        }
    }
```
So, it is unclear how this default implementation is consumed by ASP.NET pipeline...
Well... the questions is "How is the setter of `HttpContext` property consumed?"
So we have a static `AsyncLocal` instance... which is wrapper for `HttpContextHolder`
Probably `AsyncLocal` keeps scoped version of `HttpContext` but I am not sure...

Reply:
Also, I've tried this official MS DI-helper: [AddHttpContextAccessor](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/http-context?view=aspnetcore-8.0#access-httpcontext-from-custom-components) which adds singleton also:
```C#
        public static IServiceCollection AddHttpContextAccessor(this IServiceCollection services)
        {
            if (services == null)
            {
                throw new ArgumentNullException(nameof(services));
            }

            services.TryAddSingleton<IHttpContextAccessor, HttpContextAccessor>();
            return services;
        }
```
But reading the docs and logically these Singleton-design should provide scoped version of the object, right? And ASP.NET apps work correctly.
So, this design is unclear for me, but it functions in Asp.NET pipeline.

Reply:
@ggnaegi
> Are we sure it's sound in case of multiplexing too?

No, not quite sure.
What should I do here in your opinion? 'cause I am confused after this code research
I can write some tests for scenario with aggregated routes, should I? But it is a bit beyond the scope.
What I can do right now is submitting one commit with code enhancement in .

Reply:
I discovered an intriguing class: [HttpDataRepository](https://github.com/ThreeMammals/Ocelot/blob/develop/src/Ocelot/Infrastructure/RequestData/HttpDataRepository.cs) in the infrastructure, which employs the same concept of sharing data across all scoped services by storing it in `HttpContext.Items`. The dilemma then arises: should one inject the native ASP.NET service `IHttpContextAccessor`, or opt for Ocelot's own `IRequestScopedDataRepository`? 😃